lib/pure/net.nim

```diff
@@ -1477,6 +1477,43 @@ proc isIpAddress*(address_str: string): bool {.tags: [].} =
     return false
   return true
 
+proc dial*(address: string, port: Port,
+           protocol = IPPROTO_TCP, buffered = true): Socket
+           {.tags: [ReadIOEffect, WriteIOEffect].} =
+  let sockType = protocol.toSockType()
+
+  let aiList = getAddrInfo(address, port, AF_UNSPEC, sockType, protocol)
+  var success = false
+  var lastError: OSErrorCode
+  var it = aiList
+  var fd: SocketHandle
+  var domain: Domain
+  while it != nil:
+    let domainOpt = it.ai_family.toKnownDomain()
+    if domainOpt.isNone:
+      it = it.ai_next
+      continue
+    domain = domainOpt.unsafeGet()
+    let fd = newNativeSocket(domain, sockType, protocol)
+    if fd == osInvalidSocket:
+      freeAddrInfo(aiList)
+      raiseOSError(osLastError())
```

Comment:
I've fixed this, but I'm not sure if freeAddrInfo can set errno. If it can, then the same problem exists in many existing networking procedures.